In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('upstart-loan-data.csv')

In [3]:
df

Unnamed: 0 Loan ID Loan Grade  Term (months) Origination Date  \
0          15758  L17471          C             36       2015-11-04   
1          46818  L49499          B             60       2016-12-19   
2           7461  L28976          C             60       2016-03-23   
3            191    L286          C             36       2014-06-10   
4          46967  L53569          B             36       2017-02-01   
...          ...     ...        ...            ...              ...   
9995       22698  L27763          A             36       2016-03-07   
9996        6891   L5683          B             36       2015-04-06   
9997       61625  L67159          B             36       2017-05-16   
9998       15024  L12888          A             36       2015-09-10   
9999       29531  L30309          E             36       2016-04-11   

      Interest Rate    APR  Origination Fee  Loan Size  Outstanding Principal  \
0             12.56  16.94           492.00     8200.0                4733.79   
1             15.98  18.64          1400.00    25000.0               22987.76   
2             18.02  20.44           325.00     6500.0                5296.60   
3             14.15  17.80           250.00     5000.0                   0.00   
4             10.02  13.43           239.10     5000.0                4388.82   
...             ...    ...              ...        ...                    ...   
9995           7.92   9.88           436.13    15400.0                8980.34   
9996          10.93  14.51           800.00    16000.0                   0.00   
9997           9.07  12.80          1363.18    26000.0               24733.78   
9998           7.23  10.72           320.00     6400.0                   0.00   
9999          26.01  29.93          1685.00    33700.0               24462.63   

      Days Late Delinquency Status  Charged Off Amount Charged Off Date  
0            89         Delinquent                 NaN              NaN  
1             0            Current                 NaN              NaN  
2             0            Current                 NaN              NaN  
3             0            Matured                 NaN              NaN  
4             0            Current                 NaN              NaN  
...         ...                ...                 ...              ...  
9995          0            Current                 NaN              NaN  
9996          0            Matured                 NaN              NaN  
9997          0            Current                 NaN              NaN  
9998          0            Matured                 NaN              NaN  
9999         51         Delinquent                 NaN              NaN  

[10000 rows x 14 columns]

### Question 1

Drop all columns except for these... <b>['Loan Grade', 'Term (months)', 'Interest Rate', 'APR', 'Origination Fee', 'Loan Size', 'Outstanding Principal', 'Days Late', 'Delinquency Status']</b>

In [4]:
df.columns

Index(['Unnamed: 0', 'Loan ID', 'Loan Grade', 'Term (months)',
       'Origination Date', 'Interest Rate', 'APR', 'Origination Fee',
       'Loan Size', 'Outstanding Principal', 'Days Late', 'Delinquency Status',
       'Charged Off Amount', 'Charged Off Date'],
      dtype='object')

In [5]:
df.drop(['Unnamed: 0', 'Loan ID', 'Origination Date', 'Charged Off Amount', 'Charged Off Date' ], axis = 1, inplace = True)

In [6]:
df

Loan Grade  Term (months)  Interest Rate    APR  Origination Fee  \
0             C             36          12.56  16.94           492.00   
1             B             60          15.98  18.64          1400.00   
2             C             60          18.02  20.44           325.00   
3             C             36          14.15  17.80           250.00   
4             B             36          10.02  13.43           239.10   
...         ...            ...            ...    ...              ...   
9995          A             36           7.92   9.88           436.13   
9996          B             36          10.93  14.51           800.00   
9997          B             36           9.07  12.80          1363.18   
9998          A             36           7.23  10.72           320.00   
9999          E             36          26.01  29.93          1685.00   

      Loan Size  Outstanding Principal  Days Late Delinquency Status  
0        8200.0                4733.79         89         Delinquent  
1       25000.0               22987.76          0            Current  
2        6500.0                5296.60          0            Current  
3        5000.0                   0.00          0            Matured  
4        5000.0                4388.82          0            Current  
...         ...                    ...        ...                ...  
9995    15400.0                8980.34          0            Current  
9996    16000.0                   0.00          0            Matured  
9997    26000.0               24733.78          0            Current  
9998     6400.0                   0.00          0            Matured  
9999    33700.0               24462.63         51         Delinquent  

[10000 rows x 9 columns]

### Interest Rates vs. APR

You may have or you will come across terms like interest rates and APR when you go out to get loans for homes, auto and in this instance (personal loans from Upstart, a lending company), as a borrower trying to get unsecured credit. The difference between interest rate and APR in the context of refinancing or taking out a mortgage (for a home) is...

- Interest rate refers to the annual cost of a loan to a borrower and is expressed as a percentage.
- APR is the annual cost of a loan to a borrower — including fees. Like an interest rate, the APR is expressed as a percentage. Unlike an interest rate however, it includes other charges or fees such as mortgage insurance, most closing costs, discount points and loan origination fees.

But what's the difference between unsecured and secured loan? If you borrow money to buy a home, say from a bank and if you were to skip paying on that loan, guess what the bank does? It takes away your home. So the bank is making a loan based on the fact that in worst-case scenarios, they have an asset (home) they can go after. That's secured credit and hence generally carry a lower interest rate.

Unsecured credit would be like a bank lending money to you on your credit card. If you don't pay, they will cancel your credit and destroy your credit score but that is it. They can't technically come after your other assets. That's a form of unsecured credit.

### Question 2

Now that you know the difference, drop the <b>Interest Rate</b> column as well and rename the column <b>Term (months)</b> to <b>Term</b>. Also check if all attributes are in the correct format (integers, float etc). If not, change them into the right format.

In [7]:
df.drop(['Interest Rate'], axis = 1, inplace = True)

In [8]:
df.rename(columns={'Term (months)': 'Term'}, inplace=True)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Loan Grade             10000 non-null  object 
 1   Term                   10000 non-null  int64  
 2   APR                    10000 non-null  float64
 3   Origination Fee        10000 non-null  float64
 4   Loan Size              10000 non-null  float64
 5   Outstanding Principal  10000 non-null  float64
 6   Days Late              10000 non-null  int64  
 7   Delinquency Status     10000 non-null  object 
dtypes: float64(4), int64(2), object(2)
memory usage: 625.1+ KB


In [10]:
df

Loan Grade  Term    APR  Origination Fee  Loan Size  \
0             C    36  16.94           492.00     8200.0   
1             B    60  18.64          1400.00    25000.0   
2             C    60  20.44           325.00     6500.0   
3             C    36  17.80           250.00     5000.0   
4             B    36  13.43           239.10     5000.0   
...         ...   ...    ...              ...        ...   
9995          A    36   9.88           436.13    15400.0   
9996          B    36  14.51           800.00    16000.0   
9997          B    36  12.80          1363.18    26000.0   
9998          A    36  10.72           320.00     6400.0   
9999          E    36  29.93          1685.00    33700.0   

      Outstanding Principal  Days Late Delinquency Status  
0                   4733.79         89         Delinquent  
1                  22987.76          0            Current  
2                   5296.60          0            Current  
3                      0.00          0            Matured  
4                   4388.82          0            Current  
...                     ...        ...                ...  
9995                8980.34          0            Current  
9996                   0.00          0            Matured  
9997               24733.78          0            Current  
9998                   0.00          0            Matured  
9999               24462.63         51         Delinquent  

[10000 rows x 8 columns]

### Question 3

Describe in a few words what type of variables are <b>Loan Grade</b>, <b>Term</b>, <b>Days Late</b> and <b>Delinquency Status</b> and why.

In [11]:
df.loc[:,'Loan Grade'].unique()

array(['C', 'B', 'D', 'A', 'E', 'AA', 'AAA'], dtype=object)

<b>ANSWER: Loan Grade is Categorical Ordinal variable as the grade can be ordered from the best grade 'AAA', then to 'AA', 'A', 'B', 'C', 'D' and down to the worst 'E'.</b>

In [12]:
df.loc[:,'Term'].unique()

array([36, 60], dtype=int64)

<b>ANSWER: Term is Numerical Discrete variable as the values are whole positive numbers.</b>

In [13]:
df.loc[:,'Days Late'].unique()

array([ 89,   0,  15, 261,   8,   5,  88, 484, 277, 173,  13, 299,  42,
       537,  43, 440, 234,  20, 226, 122, 150, 381,  86,  10,  41, 238,
       212, 257, 255,  72,  31, 554, 110, 133,  22, 304, 430,  24, 284,
       365, 363, 187,  73, 387, 494, 319, 396, 127, 578, 193, 404, 594,
        34, 153, 170,  12, 389,   6, 442, 263, 370,  56, 520,  17, 538,
       108,  50,  21, 719,  16, 297,  61, 593,   9,   7,  74, 407,  23,
       138, 195,   4,  77, 445, 273,  11, 152, 497, 278, 414, 588, 955,
       243, 411,  52, 639,  26, 155,  49, 103, 599, 279,  28, 646, 121,
       632, 101, 585, 643,  92,  98,  27, 181,  87,  99, 390, 251,  67,
        75,  85, 463, 558, 228, 107, 685,  19, 225, 409,  46, 135, 483,
        14, 350,  45,  36, 532, 176, 358, 503, 311, 264, 120, 486, 242,
        91,  39, 289, 637, 220,  18, 426, 392, 674, 250, 435, 731, 624,
       838, 349, 560,  30, 425,  71, 324,  59, 559,  79, 662, 113,  69,
       511, 402, 783, 746, 450, 823,  96, 178, 177,  84, 166, 39

<b>ANSWER: Days Late is Numerical Continuous variable as they can be any values.</b>

In [14]:
df.loc[:,'Delinquency Status'].unique()

array(['Delinquent', 'Current', 'Matured', 'In Grace', 'Charged Off',
       'Partially Charged Off'], dtype=object)

<b>ANSWER: Deliquency Status is Categorical Nominal as they are unordered and mutually exclusive. </b>

### Question 4

Compute the average by <b>Loan Grade</b> for all columns and give a few remarks on what you observe for the <b>APR</b> and <b>Days Late</b> columns.

In [15]:
df.groupby(by = 'Loan Grade').mean()

Term        APR  Origination Fee     Loan Size  \
Loan Grade                                                        
A           40.481858  11.226493       608.045951  12282.244799   
AA          45.478060   8.464965       398.056490  13238.337182   
AAA         48.000000   6.518333       123.333333  12333.333333   
B           43.276596  14.380888       634.248265  12351.097074   
C           39.492491  17.312299       637.277255  11287.053340   
D           41.292308  21.407568       713.364571  12425.421245   
E           48.665546  27.348857       589.105403   9962.857143   

            Outstanding Principal  Days Late  
Loan Grade                                    
A                     6428.668974   6.161587  
AA                    6899.800139   2.766744  
AAA                   9117.633333   0.000000  
B                     7739.511007  10.098404  
C                     7159.271932  15.167271  
D                     6545.772066  39.327473  
E                     7470.002361  18.111765

<b>ANSWER: As you can see, the order of Loan Grade from best AAA to worst E is consistent with the ascending order of APR and mostly in ascending order of Days Late. The bad loan grade e.g. D or E are usually with higher average APR and they are also the ones with higher number of Days Late while the good Loan Grade 'A', 'AA' and 'AAA' are with relatively lower APR and smaller Days Late. </b>

### Question 5

Verify if there are any missing values in the dataset. Confirm that by applying the <b>size()</b> aggregation method on any of the columns.

In [16]:
df.groupby(by = 'Term').size().sum()

10000

In [17]:
df.groupby(by = 'APR').size().sum()

10000

In [18]:
df.groupby(by = 'Origination Fee').size().sum()

10000

In [19]:
df.groupby(by = 'Loan Size').size().sum()

10000

In [20]:
df.groupby(by = 'Outstanding Principal').size().sum()

10000

In [21]:
df.groupby(by = 'Days Late').size().sum()

10000

In [22]:
df.isnull().sum()

Loan Grade               0
Term                     0
APR                      0
Origination Fee          0
Loan Size                0
Outstanding Principal    0
Days Late                0
Delinquency Status       0
dtype: int64

In [23]:
df.notnull().sum()

Loan Grade               10000
Term                     10000
APR                      10000
Origination Fee          10000
Loan Size                10000
Outstanding Principal    10000
Days Late                10000
Delinquency Status       10000
dtype: int64

### Question 6

Display the count by <b>Loan Grade</b> of all the customers in the dataset. Also display the <b>max</b>, <b>min</b> and <b>mean</b> by <b>Loan Grade</b> of <b>Outstanding Principal</b>. Which loan grade has the least number of customers? Which loan grade has the lowest spread? Does the rating of the lowest spread customers conform to what you would expect if you were the lender?

In [24]:
df.groupby(by = 'Loan Grade').count()

Term   APR  Origination Fee  Loan Size  Outstanding Principal  \
Loan Grade                                                                  
A           2067  2067             2067       2067                   2067   
AA           433   433              433        433                    433   
AAA            6     6                6          6                      6   
B           3008  3008             3008       3008                   3008   
C           1931  1931             1931       1931                   1931   
D           1365  1365             1365       1365                   1365   
E           1190  1190             1190       1190                   1190   

            Days Late  Delinquency Status  
Loan Grade                                 
A                2067                2067  
AA                433                 433  
AAA                 6                   6  
B                3008                3008  
C                1931                1931  
D                1365                1365  
E                1190                1190

In [25]:
df.groupby(by = 'Loan Grade')['Outstanding Principal'].agg(['mean', 'max', 'min', ('Spread', lambda x : x.max() - x.min())])

mean       max  min    Spread
Loan Grade                                      
A           6428.668974  49351.47  0.0  49351.47
AA          6899.800139  49332.70  0.0  49332.70
AAA         9117.633333  21720.21  0.0  21720.21
B           7739.511007  49380.34  0.0  49380.34
C           7159.271932  50000.00  0.0  50000.00
D           6545.772066  49436.30  0.0  49436.30
E           7470.002361  50000.00  0.0  50000.00

<b>ANSWER: Loan Grade AAA has the least number of customers (6). Loan Grade AAA has the lowest spread (21720.21). Yes, the rating of the lowest spread customers conforms to what I would expect if I was the lender. They tend to be low risk borrower, with better stewardship, repaying the loan soon, hence, the outstanding principal in most cases. </b>

### Question 7

Does <b>Loan Size</b> provide any information about <b>Loan Grade</b>? Is a high risk borrower more likely to take bigger loans than a low risk borrower based on the data you have? Define in a few words what risk is to a lender? And to a borrower?

In [26]:
df.groupby(by = 'Loan Grade')['Loan Size'].agg(['mean', 'max', 'min', ('Spread', lambda x : x.max() - x.min()), 'median'])

mean      max     min   Spread   median
Loan Grade                                                 
A           12282.244799  50000.0  1000.0  49000.0  10000.0
AA          13238.337182  50000.0  1000.0  49000.0  12000.0
AAA         12333.333333  25000.0  5000.0  20000.0   9000.0
B           12351.097074  50000.0  1000.0  49000.0  10000.0
C           11287.053340  50000.0  1000.0  49000.0  10000.0
D           12425.421245  50000.0  1000.0  49000.0  10000.0
E            9962.857143  50000.0  1000.0  49000.0   7500.0

In [27]:
df.groupby(by = 'Loan Grade').apply(lambda d: d.groupby('Loan Size').size())

Loan Grade  Loan Size
A           1000.0       12
            1200.0        3
            1400.0        2
            1500.0       12
            1600.0        1
                         ..
E           40100.0       1
            42000.0       1
            44000.0       1
            47000.0       1
            50000.0      14
Length: 1258, dtype: int64

In [28]:
df_temp = df.groupby(by = 'Loan Grade').size()

In [29]:
df_temp

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
dtype: int64

In [30]:
df.loc[df['Loan Size'] > 40000,:].groupby(by = 'Loan Grade')['Loan Size'].agg([('Loan > 40K', 'size')]) 


Loan > 40K
Loan Grade            
A                   13
AA                   1
B                   23
C                   11
D                   23
E                   18

In [31]:
df.loc[ (df['Loan Size'] >= 30000) & (df['Loan Size'] < 40000),:].groupby(by = 'Loan Grade')['Loan Size'].agg([('30K <= Loan < 40K', 'size')])

30K <= Loan < 40K
Loan Grade                   
A                          72
AA                         18
B                         129
C                          65
D                          70
E                          26

In [32]:
df.loc[ (df['Loan Size'] >= 20000) & (df['Loan Size'] < 30000),:].groupby(by = 'Loan Grade')['Loan Size'].agg([('20K <= Loan < 30K', 'size')])

20K <= Loan < 30K
Loan Grade                   
A                         274
AA                         67
AAA                         2
B                         386
C                         213
D                         169
E                          89

In [33]:
df.loc[ (df['Loan Size'] >= 10000) & (df['Loan Size'] < 20000),:].groupby(by = 'Loan Grade')['Loan Size'].agg([('10K <= Loan < 20K', 'size')])

10K <= Loan < 20K
Loan Grade                   
A                         823
AA                        199
AAA                         1
B                        1108
C                         675
D                         442
E                         296

In [34]:
df.loc[ df['Loan Size'] < 10000,:].groupby(by = 'Loan Grade')['Loan Size'].agg([('Loan < 10K', 'size')])

Loan < 10K
Loan Grade            
A                  882
AA                 148
AAA                  3
B                 1350
C                  964
D                  656
E                  757

In [35]:
df_temp

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
dtype: int64

<b>ANSWER: From the statistics, there does not seem much direct obvious relationship between Loan Size and Loan Grade. Regardless of Loan Grade, most Loan sizes are less than $20000. There are bad Loan Grade customers (D, E) with Loan Size over 30K while good Loan Grade (A, AA) ones with small Loan Size (e.g. < 10K).  For Loan Size over 40K, there are more Bad Grade borrowers (D, E) than Good Grade borrowers (A, AA, AAA). For Loan Size between 30K and 40K, that relationship is not as obvious.</b>

<b>The risk to a lender is how punctual the borrow repays the interest, how long can the loan interest and principal be paid off. 
The risk to a borrower is how flunctuating the interest rate can be vary from time to time. </b>

### Question 8

Display the count by <b>Loan Grade</b> of all the customers in the dataset and call that column <b>Count</b>. Also display the <b>max</b>, <b>min</b> and <b>mean</b> by <b>Loan Grade</b> of <b>Outstanding Principal</b> and call them <b>Max Balance</b>, <b>Min Balance</b> and <b>Average Balance</b>. 

In [36]:
df.groupby(by = 'Loan Grade')['Outstanding Principal'].agg([('Count', 'count'), ('Max Balance', 'max'), ('Min Balance', 'min'), ('Average Balance', 'mean')])

Count  Max Balance  Min Balance  Average Balance
Loan Grade                                                  
A            2067     49351.47          0.0      6428.668974
AA            433     49332.70          0.0      6899.800139
AAA             6     21720.21          0.0      9117.633333
B            3008     49380.34          0.0      7739.511007
C            1931     50000.00          0.0      7159.271932
D            1365     49436.30          0.0      6545.772066
E            1190     50000.00          0.0      7470.002361

### Question 9

Write a code that takes the <b>Term</b> column and replace the values within that to <b>0</b> for <b>36</b> months and <b>1</b> for <b>60</b> months using a loop. Do the same using a lambda function. And then, using the <b>get_dummies</b> method (you might have to write additional code to drop redundant columns).

In [37]:
df.groupby(by = 'Term').size()

Term
36    7318
60    2682
dtype: int64

In [38]:
df

Loan Grade  Term    APR  Origination Fee  Loan Size  \
0             C    36  16.94           492.00     8200.0   
1             B    60  18.64          1400.00    25000.0   
2             C    60  20.44           325.00     6500.0   
3             C    36  17.80           250.00     5000.0   
4             B    36  13.43           239.10     5000.0   
...         ...   ...    ...              ...        ...   
9995          A    36   9.88           436.13    15400.0   
9996          B    36  14.51           800.00    16000.0   
9997          B    36  12.80          1363.18    26000.0   
9998          A    36  10.72           320.00     6400.0   
9999          E    36  29.93          1685.00    33700.0   

      Outstanding Principal  Days Late Delinquency Status  
0                   4733.79         89         Delinquent  
1                  22987.76          0            Current  
2                   5296.60          0            Current  
3                      0.00          0            Matured  
4                   4388.82          0            Current  
...                     ...        ...                ...  
9995                8980.34          0            Current  
9996                   0.00          0            Matured  
9997               24733.78          0            Current  
9998                   0.00          0            Matured  
9999               24462.63         51         Delinquent  

[10000 rows x 8 columns]

In [39]:
temp = df.copy()
temp.loc[temp.Term == 36, 'Term'] = 0
temp.loc[temp.Term == 60, 'Term'] = 1
temp

Loan Grade  Term    APR  Origination Fee  Loan Size  \
0             C     0  16.94           492.00     8200.0   
1             B     1  18.64          1400.00    25000.0   
2             C     1  20.44           325.00     6500.0   
3             C     0  17.80           250.00     5000.0   
4             B     0  13.43           239.10     5000.0   
...         ...   ...    ...              ...        ...   
9995          A     0   9.88           436.13    15400.0   
9996          B     0  14.51           800.00    16000.0   
9997          B     0  12.80          1363.18    26000.0   
9998          A     0  10.72           320.00     6400.0   
9999          E     0  29.93          1685.00    33700.0   

      Outstanding Principal  Days Late Delinquency Status  
0                   4733.79         89         Delinquent  
1                  22987.76          0            Current  
2                   5296.60          0            Current  
3                      0.00          0            Matured  
4                   4388.82          0            Current  
...                     ...        ...                ...  
9995                8980.34          0            Current  
9996                   0.00          0            Matured  
9997               24733.78          0            Current  
9998                   0.00          0            Matured  
9999               24462.63         51         Delinquent  

[10000 rows x 8 columns]

In [40]:
temp = df.copy()
temp.Term = temp.Term.apply(lambda x: 0 if x == 36 else 1)
temp

Loan Grade  Term    APR  Origination Fee  Loan Size  \
0             C     0  16.94           492.00     8200.0   
1             B     1  18.64          1400.00    25000.0   
2             C     1  20.44           325.00     6500.0   
3             C     0  17.80           250.00     5000.0   
4             B     0  13.43           239.10     5000.0   
...         ...   ...    ...              ...        ...   
9995          A     0   9.88           436.13    15400.0   
9996          B     0  14.51           800.00    16000.0   
9997          B     0  12.80          1363.18    26000.0   
9998          A     0  10.72           320.00     6400.0   
9999          E     0  29.93          1685.00    33700.0   

      Outstanding Principal  Days Late Delinquency Status  
0                   4733.79         89         Delinquent  
1                  22987.76          0            Current  
2                   5296.60          0            Current  
3                      0.00          0            Matured  
4                   4388.82          0            Current  
...                     ...        ...                ...  
9995                8980.34          0            Current  
9996                   0.00          0            Matured  
9997               24733.78          0            Current  
9998                   0.00          0            Matured  
9999               24462.63         51         Delinquent  

[10000 rows x 8 columns]

In [41]:
temp = df.copy()
temp = pd.get_dummies(temp, columns=['Term'])
temp

Loan Grade    APR  Origination Fee  Loan Size  Outstanding Principal  \
0             C  16.94           492.00     8200.0                4733.79   
1             B  18.64          1400.00    25000.0               22987.76   
2             C  20.44           325.00     6500.0                5296.60   
3             C  17.80           250.00     5000.0                   0.00   
4             B  13.43           239.10     5000.0                4388.82   
...         ...    ...              ...        ...                    ...   
9995          A   9.88           436.13    15400.0                8980.34   
9996          B  14.51           800.00    16000.0                   0.00   
9997          B  12.80          1363.18    26000.0               24733.78   
9998          A  10.72           320.00     6400.0                   0.00   
9999          E  29.93          1685.00    33700.0               24462.63   

      Days Late Delinquency Status  Term_36  Term_60  
0            89         Delinquent        1        0  
1             0            Current        0        1  
2             0            Current        0        1  
3             0            Matured        1        0  
4             0            Current        1        0  
...         ...                ...      ...      ...  
9995          0            Current        1        0  
9996          0            Matured        1        0  
9997          0            Current        1        0  
9998          0            Matured        1        0  
9999         51         Delinquent        1        0  

[10000 rows x 9 columns]

In [42]:
temp.drop(['Term_60'], axis = 1, inplace = True)
temp

Loan Grade    APR  Origination Fee  Loan Size  Outstanding Principal  \
0             C  16.94           492.00     8200.0                4733.79   
1             B  18.64          1400.00    25000.0               22987.76   
2             C  20.44           325.00     6500.0                5296.60   
3             C  17.80           250.00     5000.0                   0.00   
4             B  13.43           239.10     5000.0                4388.82   
...         ...    ...              ...        ...                    ...   
9995          A   9.88           436.13    15400.0                8980.34   
9996          B  14.51           800.00    16000.0                   0.00   
9997          B  12.80          1363.18    26000.0               24733.78   
9998          A  10.72           320.00     6400.0                   0.00   
9999          E  29.93          1685.00    33700.0               24462.63   

      Days Late Delinquency Status  Term_36  
0            89         Delinquent        1  
1             0            Current        0  
2             0            Current        0  
3             0            Matured        1  
4             0            Current        1  
...         ...                ...      ...  
9995          0            Current        1  
9996          0            Matured        1  
9997          0            Current        1  
9998          0            Matured        1  
9999         51         Delinquent        1  

[10000 rows x 8 columns]

### Question 10

<p>Ask an interesting question of this data set and provide a solution to answer that.</p>
<p>The goal is to help teach your fellow students all possible questions we can collectively ask of this dataset. Your question should be as clear as possible and as short as possible. Try to avoid asking questions that are too trivial or obvious.</p>

<p>This is a bonus point question. The only way not to get full points are if you do the following:
<ul>
<li>You do not perform this task.
<li>You do not include a solution. 
</ul>
</p>
<p>If the solution you provide is incorrect, you will still receive full points. But you must make an honest effort to get it right.
</p>

Write your question here.

In [43]:
df

Loan Grade  Term    APR  Origination Fee  Loan Size  \
0             C    36  16.94           492.00     8200.0   
1             B    60  18.64          1400.00    25000.0   
2             C    60  20.44           325.00     6500.0   
3             C    36  17.80           250.00     5000.0   
4             B    36  13.43           239.10     5000.0   
...         ...   ...    ...              ...        ...   
9995          A    36   9.88           436.13    15400.0   
9996          B    36  14.51           800.00    16000.0   
9997          B    36  12.80          1363.18    26000.0   
9998          A    36  10.72           320.00     6400.0   
9999          E    36  29.93          1685.00    33700.0   

      Outstanding Principal  Days Late Delinquency Status  
0                   4733.79         89         Delinquent  
1                  22987.76          0            Current  
2                   5296.60          0            Current  
3                      0.00          0            Matured  
4                   4388.82          0            Current  
...                     ...        ...                ...  
9995                8980.34          0            Current  
9996                   0.00          0            Matured  
9997               24733.78          0            Current  
9998                   0.00          0            Matured  
9999               24462.63         51         Delinquent  

[10000 rows x 8 columns]

<b>ANSWER: Do the loan with bad loan grade always go delinquent?</b>

In [44]:
df.groupby(by = 'Loan Grade').size()

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
dtype: int64

In [45]:
temp = df.copy()
temp = pd.get_dummies(temp, columns=['Delinquency Status'])
temp

Loan Grade  Term    APR  Origination Fee  Loan Size  \
0             C    36  16.94           492.00     8200.0   
1             B    60  18.64          1400.00    25000.0   
2             C    60  20.44           325.00     6500.0   
3             C    36  17.80           250.00     5000.0   
4             B    36  13.43           239.10     5000.0   
...         ...   ...    ...              ...        ...   
9995          A    36   9.88           436.13    15400.0   
9996          B    36  14.51           800.00    16000.0   
9997          B    36  12.80          1363.18    26000.0   
9998          A    36  10.72           320.00     6400.0   
9999          E    36  29.93          1685.00    33700.0   

      Outstanding Principal  Days Late  Delinquency Status_Charged Off  \
0                   4733.79         89                               0   
1                  22987.76          0                               0   
2                   5296.60          0                               0   
3                      0.00          0                               0   
4                   4388.82          0                               0   
...                     ...        ...                             ...   
9995                8980.34          0                               0   
9996                   0.00          0                               0   
9997               24733.78          0                               0   
9998                   0.00          0                               0   
9999               24462.63         51                               0   

      Delinquency Status_Current  Delinquency Status_Delinquent  \
0                              0                              1   
1                              1                              0   
2                              1                              0   
3                              0                              0   
4                              1                              0   
...                          ...                            ...   
9995                           1                              0   
9996                           0                              0   
9997                           1                              0   
9998                           0                              0   
9999                           0                              1   

      Delinquency Status_In Grace  Delinquency Status_Matured  \
0                               0                           0   
1                               0                           0   
2                               0                           0   
3                               0                           1   
4                               0                           0   
...                           ...                         ...   
9995                            0                           0   
9996                            0                           1   
9997                            0                           0   
9998                            0                           1   
9999                            0                           0   

      Delinquency Status_Partially Charged Off  
0                                            0  
1                                            0  
2                                            0  
3                                            0  
4                                            0  
...                                        ...  
9995                                         0  
9996                                         0  
9997                                         0  
9998                                         0  
9999                                         0  

[10000 rows x 13 columns]

In [46]:
temp.groupby(by = 'Loan Grade')['Delinquency Status_Delinquent'].agg([('Delinquent Count','sum'), ('Total Loan', 'count'), ('Delinquent %', lambda x : x.sum() / x.count() * 100)])

C:\Users\EW\anaconda3\lib\site-packages\ipykernel_launcher.py:1: RuntimeWarning: invalid value encountered in longlong_scalars
  """Entry point for launching an IPython kernel.


Delinquent Count  Total Loan  Delinquent %
Loan Grade                                            
A                         19        2067      0.919207
AA                         2         433      0.461894
AAA                        0           6      0.000000
B                         60        3008      1.994681
C                         55        1931      2.848265
D                         44        1365      3.223443
E                         37        1190      3.109244

In [47]:
temp.groupby(by = 'Loan Grade').apply(lambda d: d.loc[(d.APR == d.APR.max()) | (d.APR == d.APR.min()), ['APR', 'Loan Size', 'Days Late']])

APR  Loan Size  Days Late
Loan Grade                                  
A          1391  15.22    10000.0          0
           6504   6.98    12100.0          0
AA         4358   4.30    17000.0          0
           8966  12.31     4300.0          0
AAA        4676   7.19    20000.0          0
           5700   6.14     6000.0          0
B          2438  21.82    50000.0          0
           4968   9.73     3000.0          0
C          7995  25.41    48200.0          0
           8004  13.27    10000.0          0
D          223   29.99    18600.0          0
           1280  29.99     3800.0          0
           2762  29.99    19600.0          0
           4525  29.99     9800.0          0
           5279  29.99     4400.0          0
           5299  29.99     3300.0          0
           8898  15.96     7500.0          0
           9054  29.99     6200.0          0
           9057  29.99    23800.0          0
E          44    29.99     1400.0          0
           740   29.99     4800.0        153
           2106  29.99     5200.0          0
           3284  29.99     9200.0          0
           3513  29.99     1000.0          0
           3881  29.99     1500.0          0
           4145  29.99     5700.0        292
           4731  29.99     7000.0        152
           5618  29.99     1400.0          0
           5761  29.99     9400.0          0
           6123  29.99    14100.0          0
           7732  18.76     3600.0          0
           8150  29.99     5300.0          0
           8200  29.99    12300.0          0
           8721  29.99    13600.0          0
           9188  29.99     8200.0          0

<b>ANSWER: From the statistics, better the loan grade, lower the delinqency rate. Bad Loan Grade (D, E) usually results in higher rate of delinquency while those of the good Loan Grade (AAA,AA,A) are lower. That may explain why bank usually charge higher APR to bad Loan Grade ones to protect their interest and lower their risk in lending.</b>